ssl/t1_enc.c

```diff
@@ -316,6 +325,64 @@ int tls1_change_cipher_state(SSL *s, int which)
                  ERR_R_INTERNAL_ERROR);
         goto err;
     }
+#ifndef OPENSSL_NO_KTLS
+    if (s->compress || SSL_TREAT_AS_TLS13(s))
+        goto skip_ktls;
+
+    if ((which & SSL3_CC_READ) ||
+        ((which & SSL3_CC_WRITE) && (s->mode & SSL_MODE_NO_KTLS_TX)))
+        goto skip_ktls;
+
+    /* ktls supports only the maximum fragment size */
+    if (ssl_get_max_send_fragment(s) != SSL3_RT_MAX_PLAIN_LENGTH)
+        goto skip_ktls;
+
+    /* check that cipher is AES_GCM_128 */
+    if (EVP_CIPHER_nid(c) != NID_aes_128_gcm
+        || EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE
+        || EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE)
+        goto skip_ktls;
+
+    /* check version is 1.2 */
+    if (s->version != TLS1_2_VERSION)
+        goto skip_ktls;
+
+    wbio = s->wbio;
+    if (!ossl_assert(wbio != NULL)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS1_CHANGE_CIPHER_STATE,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    /* All future data will get encrypted by ktls. Flush the BIO or skip ktls */
+    if (BIO_flush(wbio) <= 0)
+        goto skip_ktls;
+
+    /* ktls doesn't support renegotiation */
+    if (BIO_get_ktls(s->wbio))
+        goto err;
+
+    /* ktls works with user provided buffers directly */
+    ssl3_release_write_buffer(s);
```

Comment:
Should we wait for success of BIO_set_ktls() before releasing the write buffer? If the underlying BIO doesn't support it, or it fails for some other reason we have needlessly released it.

Reply:
What do you mean by ~wait for success~?

It is successful if we reach `ssl3_release_write_buffer(s);`

Reply:
BIO_set_ktls() might still fail at this point, e.g. because we're using a mem BIO, or we've compiled in the dummy KTLS code because the kernel doesn't support it or it fails for some other reason.

Reply:
Should be fixed now.